packages/flutter/lib/src/widgets/image.dart

```diff
@@ -459,6 +543,203 @@ class Image extends StatefulWidget {
   /// The image to display.
   final ImageProvider image;
 
+  /// A builder function responsible for creating the widget that represents
+  /// this image.
+  ///
+  /// If this is null, this widget will display an image that is painted as
+  /// soon as the first frame is available (and will appear to "pop" in if it
+  /// becomes available asynchronously). Callers might use this builder to
+  /// add effects to the image (such as fading the image in when it becomes
+  /// available) or to display a placeholder widget while the image is loading.
+  ///
+  /// To have finer-grained control over the way that an image's loading
+  /// progress is communicated to the user, see [loadingBuilder].
+  ///
+  /// ## Chaining with [loadingBuilder]
+  ///
+  /// If a [loadingBuilder] has _also_ been specified for an image, the two
+  /// builders will be chained together, and the _result_ of this builder will
+  /// be passed as the `child` argument to the [loadingBuilder]. For example,
+  /// consider the following builders used in conjunction:
+  ///
+  /// ```dart
+  /// Image(
+  ///   ...
+  ///   frameBuilder: (BuildContext context, Widget child, int frame, bool wasSynchronouslyLoaded) {
+  ///     return Padding(
+  ///       padding: EdgeInsets.all(8.0),
+  ///       child: child,
+  ///     );
+  ///   },
+  ///   loadingBuilder: (BuildContext context, Widget child, ImageChunkEvent loadingProgress) {
+  ///     return Center(child: child);
+  ///   },
+  /// )
+  /// ```
+  ///
+  /// In this example, while the image is loading, the widget hierarchy will
+  /// contain the following:
+  ///
+  /// ```dart
+  /// Center(
+  ///   Padding(
+  ///     padding: EdgeInsets.all(8.0),
+  ///     child: <image>,
+  ///   ),
+  /// )
+  /// ```
+  ///
+  /// Then, in the same example, once the image is done loading, the
+  /// [loadingBuilder] will no longer be consulted, and the widget hierarchy
+  /// will contain the following:
+  ///
+  /// ```dart
+  /// Padding(
+  ///   padding: EdgeInsets.all(8.0),
+  ///   child: <image>,
+  /// )
```

Comment:
that will cause the RenderImage to be disposed and a new one created, which is unfortunate... do we really want that?

Reply:
Good point.

We could consult the loading builder even when there's no chunk event to report, with a null ImageChunkEvent.  That would enable them to return a consistent widget subtree.

(the caller could also solve it using global keys, but I acknowledge that's a horrible answer)

Reply:
I updated the API to always consult `loadingBuilder` and to pass a null `loadingProgress` argument when the image hasn't started loading yet or when it's fully loaded.  Callers can then choose to return `child` directly if `loadingBuilder` is null if they wish, at the cost of disposing and creating a new `RenderImage`.